refactored the smart ptr class macros

wjwwood opened the pull request:
The macros are now more modular and there they include weak pointer and unique pointer definitions.

On classes which are not copy constructible we cannot make a `make_unique` definition, so there are two main entry point macros now, `RCLCPP_SMART_PTR_DEFINITIONS` and `RCLCPP_SMART_PTR_DEFINITIONS_NOT_COPYABLE`. Where the latter is used on classes which are not copyable. I worked with @esteve to come up with an automatic solution to this using `enable_if` or something similar, but it doesn't seem to be possible with only one macro inside the class definition. So I made two macros instead. It will not compile if you use the wrong macro, so that should be ok.

wjwwood commented:
These changes were pull out of https://github.com/ros2/rclcpp/pull/73, so I'll rebase that one once this is merged. For review.

wjwwood commented:
CI:

- Linux: http://ci.ros2.org/job/ros2_batch_ci_linux/188/
- OS X: (something is going on, looking into it)

esteve commented on rclcpp/include/rclcpp/macros.hpp:
```diff
@@ -23,14 +23,33 @@
+  RCLCPP_WEAK_PTR_DEFINITIONS(Class) \
+  __RCLCPP_UNIQUE_PTR_TYPEDEF(Class)
+
+#define __RCLCPP_SHARED_PTR_TYPEDEF(Class) typedef std::shared_ptr<Class> SharedPtr;
```
Just an aesthetic suggestion. Given that we're redoing the macros, we can start using `using` instead of `typedef`:

```c++
#define __RCLCPP_SHARED_PTR_TYPEDEF(Class) using SharedPtr = std::shared_ptr<Class>;
```

this would make the code more uniform if any of the aliases require a template, since `typedef` can't be templated.

esteve commented:
+1

dirk-thomas commented:
+1 for the proposed changes (ignoring the result of the CI job)

wjwwood commented:
Ok, I fixed the bug (it was caused by selecting partial diffs to commit out of the previous pr) and I switched to `using X = ...` rather than typedefs.

New Linux job: http://ci.ros2.org/job/ros2_batch_ci_linux/190/

OS X is still broken for some reason.

wjwwood commented:
Ok, here is the OS X one: http://ci.ros2.org/job/ros2_batch_ci_osx/125/

esteve commented on rclcpp/include/rclcpp/macros.hpp:
```diff
@@ -23,14 +23,33 @@
+  RCLCPP_WEAK_PTR_DEFINITIONS(Class) \
+  __RCLCPP_UNIQUE_PTR_TYPEDEF(Class)
+
+#define __RCLCPP_SHARED_PTR_TYPEDEF(Class) using SharedPtr = std::shared_ptr<Class>;
```
I'd rename name this macro to `__RCLCPP_SHARED_PTR_ALIAS` (or something similar), given that it's no longer a `typedef`

wjwwood replied:
Good point, done as a fixup commit.

wjwwood commented:
A working OS X job: http://ci.ros2.org/job/ros2_batch_ci_osx/126/

wjwwood commented:
I'm going to merge, @tfoote @jacquelinekay you can comment after the fact if you like.
